contracts/starknet/teleport_router.cairo

```diff
@@ -0,0 +1,297 @@
+    let prev_gateway = _gateways.read(domain)
+
+    # if(prevGateway == address(0)) {
+    if prev_gateway == 0:
+        # if(data != address(0)) allDomains.add(domain);
+        let (is_data_zero) = is_not_zero(data)
+        if is_data_zero == 0:
+            # allDomains.add(domain); TODO
+        end
+    else:
+        # if(data == address(0)) allDomains.remove(domain);
+        if data == 0:
+            # allDomains.remove(domain); TODO
+        end
+    end
+
+    # gateways[domain] = data;
+    _gateways.write(domain, data)
+
+    # else revert("TeleportRouter/file-unrecognized-param");
+    with_attr error_message("TeleportRouter/file-unrecognized-param"):
+        assert what = 'gateway'
+    end
+
+    # emit File(what, domain, data);
+    File_ilk.emit(ilk, what, data)
+
+    return ()
+end
+
+# function file(bytes32 what, address data) external auth {
+@external
+func file{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    what : felt, data : Uint256
+):
+    auth()
+
+    # if (what == "parent") parent = data;
+    # else revert("TeleportRouter/file-unrecognized-param");
+    with_attr error_message("TeleportRouter/file-unrecognized-param"):
+        assert what = 'parent'
+    end
+    _parent.write(data)
+
+    # emit File(what, data);
+    File.emit(what, data)
+
+    return ()
+end
+
+# function numDomains() external view returns (uint256) {
+@view
+func numDomains{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}() -> (
+    res : felt
+):
+    # return allDomains.length(); TODO
+end
+
+# function domainAt(uint256 index) external view returns (bytes32) {
+@view
+func domainAt{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(index : felt):
+    # return allDomains.at(index); TODO
+end
+
+# function hasDomain(bytes32 domain) external view returns (bool) {
+@view
+func hasDomain{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(domain : felt):
+    # return allDomains.contains(domain); TODO
+end
+
+# function registerMint(TeleportGUID calldata teleportGUID) external {
+@external
+func registerMint{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    teleportGUID : TeleportGUID
+):
+    # require(msg.sender == parent || msg.sender == gateways[teleportGUID.sourceDomain], "TeleportRouter/sender-not-gateway");
+    with_attr error_message("TeleportRouter/sender-not-gateway"):
+        let (caller) = get_caller_address()
+        let (parent_eq) = is_not_zero(caller - parent)
+        let (source_gateway) = _gateways.read(teleportGUID.sourceDomain)
+        let (gateway_eq) = is_not_zero(caller - source_gateway)
+        assert_either(parent_eq, gateway_eq)
+    end
+
+    # address gateway = gateways[teleportGUID.targetDomain];
+    let gateway = _gateways.read(teleportGUID.targetDomain)
+
+    # if (gateway == address(0)) gateway = parent;
+    if gateway == 0:
+        _gateway.write(parent)
```

Comment:
Storage should not be updated here.

Reply:
replaced by a `get_gateway` function for conditional assignment it's the best way I know of